qiskit/pulse/library/symbolic_pulses.py

```diff
@@ -0,0 +1,810 @@
+            PulseError: When expression for pulse envelope is not assigned.
+        """
+        if self.is_parameterized():
+            raise PulseError("Unassigned parameter exists. All parameters must be assigned.")
+
+        if self.envelope is None:
+            raise PulseError("Pulse envelope expression is not assigned.")
+
+        fargs = _get_expression_args(self.envelope, self.parameters)
+        return Waveform(samples=self._envelope_lam(*fargs), name=self.name)
+
+    def validate_parameters(self) -> None:
+        """Validate parameters.
+
+        Raises:
+            PulseError: If the parameters passed are not valid.
+        """
+        if self.is_parameterized():
+            return
+
+        if self.constraints is not None:
+            fargs = _get_expression_args(self.constraints, self.parameters)
+            if not bool(self._constraints_lam(*fargs)):
+                param_repr = ", ".join(f"{p}={v}" for p, v in self.parameters.items())
+                const_repr = str(self.constraints)
+                raise PulseError(
+                    f"Assigned parameters {param_repr} violate following constraint: {const_repr}."
+                )
+
+        if self.limit_amplitude:
+            if self.eval_conditions is not None:
+                fargs = _get_expression_args(self.eval_conditions, self.parameters)
+                check_full_waveform = bool(self._eval_conditions_lam(*fargs))
+            else:
+                check_full_waveform = True
+
+            if check_full_waveform:
+                # Check full waveform only when the condition is satisified or
+                # evaluation condition is not provided.
+                # This operation is slower due to overhead of 'get_waveform'.
+                if not _is_amplitude_valid(self):
+                    param_repr = ", ".join(f"{p}={v}" for p, v in self.parameters.items())
+                    raise PulseError(
+                        f"Maximum pulse amplitude norm exceeds 1.0 with parameters {param_repr}."
+                        "This can be overruled by setting Pulse.limit_amplitude."
+                    )
+
+    def is_parameterized(self) -> bool:
+        """Return True iff the instruction is parameterized."""
+        return any(isinstance(val, ParameterExpression) for val in self.parameters.values())
+
+    @property
+    def parameters(self) -> Dict[str, Any]:
+        params = {"duration": self.duration}
+        params.update(self._params)
+        return params
+
+    def __eq__(self, other: "SymbolicPulse") -> bool:
+
+        if not isinstance(other, SymbolicPulse):
+            return NotImplemented
+
+        if self.envelope != other.envelope:
+            return False
+
+        if self.parameters != other.parameters:
+            return False
+
+        return True
+
+    def __hash__(self) -> int:
+        return hash((self._pulse_type, self.duration, *tuple(self._params.items())))
```

Comment:
I see a potential issue here. `__eq__` and `__hash__` are implemented differently. I.e. I can hash two `SymbolicPulse`s where `__eq__` is True and compare the hashes and get False. Should we aim to have `__hash__` and `__eq__` preserve symbolic pulse equality?

Reply:
Yes, `__eq__` and `__hash__` are a bit loose here right now. The main reason `__hash__` was implemented was so that `_is_valid_amplitude()` could use `lru_cache` to skip re-validating the same pulse parameters multiple times. @nkanazawa1989 said this was a concern because code often regenerates the pulses with the same data so slow validation would add a lot of overhead (and we made the validation slow in some cases to be more efficient / accurate in other ways -- if it is not easy to write down the conditions for a valid  pulse easily, like with DRAG, we fall back to full waveform evaluation).

`__hash__` is a bit weird for `SymbolicPulse` because it is a mutable object. Perhaps the `lru_cache` should be pushed down into a function that takes `envelope` and `parameters` as arguments and  `__hash__` should be dropped (unless there are other likely needs for hashing pulses?). If we want to keep `__hash__`, I think we should consider making `SymbolicPulse` less mutable -- maybe the expressions like  `envelope` can remain optional but can not be changed after `__init__`. Also, could assigning parameters generate a new object instead of mutating `_params` within the same object? If not, maybe `__hash__` should raise an exception if the parameters are not all assigned.

`__eq__` is also a bit weird. What is the use case for comparing if pulses are equal? The current form will confirm if two pulses will produce the same waveform but not that they have the same display name or validity conditions. For that, we would want to compare constraints, eval_conditions, limit_amplitude, and pulse_type as well.

Reply:
Thanks, you're right @eggerdj . This is due to missing `_pulse_type` equality which was removed after this conversation https://github.com/Qiskit/qiskit-terra/pull/7821#discussion_r887229045. I readded `_pulse_type` check to `__eq__`. I believe equality and hashing are [not a necessary and sufficient conditions](https://eng.lyft.com/hashing-and-equality-in-python-2ea8c738fb9d). So it is allowed to return the same hash for different instances.  

Note that `__hash__` is the abstract method so we must implement it. This is because, we need to investigate the pulse hash to append a unique string after pulse name for the pulse qobj serialization, i.e. specifically, to create `pulse_library` (dict) entry. This is created for `Waveform` s, and we do name-based data compression (to enjoy O(1) lookup of already serialized pulse) otherwise it drastically increases the payload size or computational overhead (i.e. in `Play(Waveform, Channel)` waveform data is separately stored in qobj) . `Waveform` must be immutable (since it doesn't take parameters) and we only had this kind of pulse when we first designed the framework, but `ParametricPulse` comes later. So the pulse hashing is no longer a correct implementation from software viewpoint. Note that these pulses are basically immutable, but I write the pulse `ParameterManager` class with a visitor pattern that is allowed to assign parameters mutably for better performance with `inplace=True`.

This is bit outside the scope of the PR. So I just turn expressions into protected. Proper immutability of pulse must be carefully considered in follow-up. See https://github.com/Qiskit/qiskit-terra/pull/7821/commits/3b2c5716e671c5a6069e87528b00ca8bd2d68c78